ttnn/cpp/ttnn/operations/pool/generic/device/pool_multi_core_program_factory.cpp

```diff
@@ -1,19 +1,97 @@
-// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
-//
+// SPDX-FileCopyrightText: © 2024 Tenstorrent AI ULC
 // SPDX-License-Identifier: Apache-2.0
 
 #include "pool_op.hpp"
 #include "tt-metalium/circular_buffer.hpp"
 #include "tt-metalium/circular_buffer_config.hpp"
 #include "ttnn/operations/cb_utils.hpp"
 #include "ttnn/operations/pool/pool_utils.hpp"
+#include "tt-metalium/host_buffer.hpp"
+#include "tt-metalium/buffer.hpp"
+#include <vector>
+#include "ttnn/tensor/storage.hpp"
 #include <tt-metalium/hal.hpp>
+#include <algorithm>
 
 namespace ttnn::operations::pool {
 /**
  * Generic pool implementation that uses the new sliding window infrastructure.
  */
 
+// This function creates a scalar config tensor for the AvgPool2D operation. It is entirely made of
+// a vector of ScalarInfo structs, which are used to configure the pooling operation. The config tensor
+// is filled with out_nhw_per_core number of ScalarInfos for each core and then sharded across the cores.
+// Since we don't usually have that many different scalars, we fill the rest of the config tensor with 0s.
+static Tensor create_scalar_config_tensor(
+    AvgPoolConfig config, uint32_t in_n, uint32_t num_shards_c, CoreRangeSet all_cores) {
+    auto ranges = all_cores.ranges();
+
+    std::vector<uint32_t> config_vector;
+
+    uint32_t output_stick_w = 0;
+    uint32_t output_stick_h = 0;
+    uint32_t output_stick_c = 0;
+    uint32_t output_stick_n = 0;
+
+    uint32_t max_scalars_cnt = 0;
+    std::vector<std::vector<ScalarInfo>> scalars_per_core = {};
```

Comment:
you can add logic to not repeat this process for cores that have exact same setup (like every core for width sharding) or at least a todo for that.